File: 2023/day14/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>

void tiltNorth(std::vector<std::string>& grid) {
    int rows = static_cast<int>(grid.size());
    int cols = static_cast<int>(grid[0].size());

    for (int col = 0; col < cols; ++col) {
        int writePos = 0;
        for (int row = 0; row < rows; ++row) {
            if (grid[row][col] == '#') {
                writePos = row + 1;
            } else if (grid[row][col] == 'O') {
                grid[row][col] = '.';
                grid[writePos][col] = 'O';
                ++writePos;
            }
        }
    }
}

void tiltSouth(std::vector<std::string>& grid) {
    int rows = static_cast<int>(grid.size());
    int cols = static_cast<int>(grid[0].size());

    for (int col = 0; col < cols; ++col) {
        int writePos = rows - 1;
        for (int row = rows - 1; row >= 0; --row) {
            if (grid[row][col] == '#') {
                writePos = row - 1;
            } else if (grid[row][col] == 'O') {
                grid[row][col] = '.';
                grid[writePos][col] = 'O';
                --writePos;
            }
        }
    }
}

void tiltWest(std::vector<std::string>& grid) {
    int rows = static_cast<int>(grid.size());
    int cols = static_cast<int>(grid[0].size());

    for (int row = 0; row < rows; ++row) {
        int writePos = 0;
        for (int col = 0; col < cols; ++col) {
            if (grid[row][col] == '#') {
                writePos = col + 1;
            } else if (grid[row][col] == 'O') {
                grid[row][col] = '.';
                grid[row][writePos] = 'O';
                ++writePos;
            }
        }
    }
}

void tiltEast(std::vector<std::string>& grid) {
    int rows = static_cast<int>(grid.size());
    int cols = static_cast<int>(grid[0].size());

    for (int row = 0; row < rows; ++row) {
        int writePos = cols - 1;
        for (int col = cols - 1; col >= 0; --col) {
            if (grid[row][col] == '#') {
                writePos = col - 1;
            } else if (grid[row][col] == 'O') {
                grid[row][col] = '.';
                grid[row][writePos] = 'O';
                --writePos;
            }
        }
    }
}

void spinCycle(std::vector<std::string>& grid) {
    tiltNorth(grid);
    tiltWest(grid);
    tiltSouth(grid);
    tiltEast(grid);
}

long long calculateLoad(const std::vector<std::string>& grid) {
    long long load = 0;
    int rows = static_cast<int>(grid.size());

    for (int row = 0; row < rows; ++row) {
        for (char c : grid[row]) {
            if (c == 'O') {
                load += rows - row;
            }
        }
    }
    return load;
}

std::string gridToString(const std::vector<std::string>& grid) {
    std::string result;
    for (const auto& row : grid) {
        result += row;
    }
    return result;
}

long long part1(std::vector<std::string> grid) {
    tiltNorth(grid);
    return calculateLoad(grid);
}

long long part2(std::vector<std::string> grid) {
    std::unordered_map<std::string, long long> seen;
    const long long targetCycles = 1000000000;

    for (long long cycle = 0; cycle < targetCycles; ++cycle) {
        std::string key = gridToString(grid);

        auto it = seen.find(key);
        if (it != seen.end()) {
            long long cycleStart = it->second;
            long long cycleLength = cycle - cycleStart;
            long long remaining = (targetCycles - cycle) % cycleLength;

            for (long long i = 0; i < remaining; ++i) {
                spinCycle(grid);
            }
            return calculateLoad(grid);
        }

        seen[key] = cycle;
        spinCycle(grid);
    }

    return calculateLoad(grid);
}

int main() {
    std::ifstream file("../input.txt");
    if (!file.is_open()) {
        std::cerr << "Error: Could not open input file" << std::endl;
        return 1;
    }

    std::vector<std::string> grid;
    std::string line;
    while (std::getline(file, line)) {
        if (!line.empty()) {
            grid.push_back(line);
        }
    }
    file.close();

    std::cout << "Part 1: " << part1(grid) << std::endl;
    std::cout << "Part 2: " << part2(grid) << std::endl;

    return 0;
}

File: 2023/day14/php/solution.php
<?php
declare(strict_types=1);

/**
 * Parse input into a 2D grid.
 *
 * @param string $text
 * @return array<int, array<int, string>>
 */
function parseInput(string $text): array
{
    $lines = explode("\n", trim($text));
    $grid = [];
    foreach ($lines as $line) {
        $grid[] = str_split($line);
    }
    return $grid;
}

/**
 * Tilt the grid north, moving all round rocks up.
 *
 * @param array<int, array<int, string>> $grid
 * @return void
 */
function tiltNorth(array &$grid): void
{
    $rows = count($grid);
    $cols = count($grid[0]);

    for ($col = 0; $col < $cols; $col++) {
        $writePos = 0;
        for ($row = 0; $row < $rows; $row++) {
            if ($grid[$row][$col] === '#') {
                $writePos = $row + 1;
            } elseif ($grid[$row][$col] === 'O') {
                $grid[$row][$col] = '.';
                $grid[$writePos][$col] = 'O';
                $writePos++;
            }
        }
    }
}

/**
 * Tilt the grid south, moving all round rocks down.
 *
 * @param array<int, array<int, string>> $grid
 * @return void
 */
function tiltSouth(array &$grid): void
{
    $rows = count($grid);
    $cols = count($grid[0]);

    for ($col = 0; $col < $cols; $col++) {
        $writePos = $rows - 1;
        for ($row = $rows - 1; $row >= 0; $row--) {
            if ($grid[$row][$col] === '#') {
                $writePos = $row - 1;
            } elseif ($grid[$row][$col] === 'O') {
                $grid[$row][$col] = '.';
                $grid[$writePos][$col] = 'O';
                $writePos--;
            }
        }
    }
}

/**
 * Tilt the grid west, moving all round rocks left.
 *
 * @param array<int, array<int, string>> $grid
 * @return void
 */
function tiltWest(array &$grid): void
{
    $rows = count($grid);
    $cols = count($grid[0]);

    for ($row = 0; $row < $rows; $row++) {
        $writePos = 0;
        for ($col = 0; $col < $cols; $col++) {
            if ($grid[$row][$col] === '#') {
                $writePos = $col + 1;
            } elseif ($grid[$row][$col] === 'O') {
                $grid[$row][$col] = '.';
                $grid[$row][$writePos] = 'O';
                $writePos++;
            }
        }
    }
}

/**
 * Tilt the grid east, moving all round rocks right.
 *
 * @param array<int, array<int, string>> $grid
 * @return void
 */
function tiltEast(array &$grid): void
{
    $rows = count($grid);
    $cols = count($grid[0]);

    for ($row = 0; $row < $rows; $row++) {
        $writePos = $cols - 1;
        for ($col = $cols - 1; $col >= 0; $col--) {
            if ($grid[$row][$col] === '#') {
                $writePos = $col - 1;
            } elseif ($grid[$row][$col] === 'O') {
                $grid[$row][$col] = '.';
                $grid[$row][$writePos] = 'O';
                $writePos--;
            }
        }
    }
}

/**
 * Perform one spin cycle: N, W, S, E.
 *
 * @param array<int, array<int, string>> $grid
 * @return void
 */
function spinCycle(array &$grid): void
{
    tiltNorth($grid);
    tiltWest($grid);
    tiltSouth($grid);
    tiltEast($grid);
}

/**
 * Convert grid to a string for cycle detection.
 *
 * @param array<int, array<int, string>> $grid
 * @return string
 */
function gridToString(array $grid): string
{
    $result = '';
    foreach ($grid as $row) {
        $result .= implode('', $row) . "\n";
    }
    return $result;
}

/**
 * Calculate total load on north support beams.
 *
 * @param array<int, array<int, string>> $grid
 * @return int
 */
function calculateLoad(array $grid): int
{
    $rows = count($grid);
    $total = 0;

    for ($row = 0; $row < $rows; $row++) {
        foreach ($grid[$row] as $cell) {
            if ($cell === 'O') {
                $total += $rows - $row;
            }
        }
    }

    return $total;
}

/**
 * Part 1: Tilt north and calculate load.
 *
 * @param array<int, array<int, string>> $grid
 * @return int
 */
function part1(array $grid): int
{
    tiltNorth($grid);
    return calculateLoad($grid);
}

/**
 * Part 2: Run 1 billion spin cycles and calculate load.
 *
 * @param array<int, array<int, string>> $grid
 * @return int
 */
function part2(array $grid): int
{
    $target = 1_000_000_000;
    $seen = [];
    $cycleNum = 0;

    while ($cycleNum < $target) {
        $state = gridToString($grid);

        if (isset($seen[$state])) {
            $cycleStart = $seen[$state];
            $cycleLength = $cycleNum - $cycleStart;
            $remaining = ($target - $cycleNum) % $cycleLength;

            for ($i = 0; $i < $remaining; $i++) {
                spinCycle($grid);
            }

            return calculateLoad($grid);
        }

        $seen[$state] = $cycleNum;
        spinCycle($grid);
        $cycleNum++;
    }

    return calculateLoad($grid);
}

/**
 * Main entry point.
 */
function main(): void
{
    $inputFile = __DIR__ . '/../input.txt';
    $text = file_get_contents($inputFile);

    if ($text === false) {
        fwrite(STDERR, "Error: Could not read input file\n");
        exit(1);
    }

    $grid = parseInput($text);

    echo "Part 1: " . part1($grid) . "\n";
    echo "Part 2: " . part2($grid) . "\n";
}

main();
